common/archiver/provider/provider.go

```diff
@@ -0,0 +1,110 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package provider
+
+import (
+	"errors"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/archiver"
+	"github.com/uber/cadence/common/archiver/filestore"
+	"github.com/uber/cadence/common/log/tag"
+)
+
+var (
+	// ErrUnknownScheme is the error for unknown archiver scheme
+	ErrUnknownScheme = errors.New("unknown archiver scheme")
+	// ErrUnknownServiceName is the error for unknown cadence service name
+	ErrUnknownServiceName = errors.New("unknown service name")
+)
+
+type (
+	// Provider returns history or visiblity archiver based on the scheme and serviceName.
+	// The archiver for each combination of scheme and serviceName will be created only once and cached.
+	Provider interface {
+		GetHistoryArchiver(scheme string, serviceName string) (archiver.HistoryArchiver, error)
+		GetVisibilityArchiver(scheme string, serviceName string) (archiver.VisibilityArchiver, error)
+	}
+
+	// HistoryArchiverConfigs contain config for all implementations of the HistoryArchiver interface
+	HistoryArchiverConfigs struct {
+		FileStore *filestore.HistoryArchiverConfig
+	}
+
+	// VisibilityArchiverConfigs contain config for all implementations of the VisibilityArchiver interface
+	VisibilityArchiverConfigs struct {
+	}
+
+	provider struct {
+		container                 *archiver.BootstrapContainer
+		historyArchiverConfigs    *HistoryArchiverConfigs
+		visibilityArchiverConfigs *VisibilityArchiverConfigs
+
+		historyArchivers    map[string]archiver.HistoryArchiver
+		visibilityArchivers map[string]archiver.VisibilityArchiver
+	}
+)
+
+// NewArchiverProvider returns a new Archiver provider
+func NewArchiverProvider(
+	container *archiver.BootstrapContainer,
+	historyArchiverConfigs *HistoryArchiverConfigs,
+	visibilityArchiverConfigs *VisibilityArchiverConfigs,
+) Provider {
+	return &provider{
+		container:                 container,
+		historyArchiverConfigs:    historyArchiverConfigs,
+		visibilityArchiverConfigs: visibilityArchiverConfigs,
+	}
+}
+
+func (p *provider) GetHistoryArchiver(scheme, serviceName string) (archiver.HistoryArchiver, error) {
+	key := p.getArchiverKey(scheme, serviceName)
+	if archiver, ok := p.historyArchivers[key]; ok {
+		return archiver, nil
+	}
+
+	var componentTag tag.Tag
+	switch serviceName {
```

Comment:
This looks weird. This method should not be aware of where it is called from but having service name tied to component tag makes this code aware of where it's being called from. I see a few ways around this but for now lets do a simple thing and just not include component tag.

Reply:
As discussed, added ...tag.Tag as a function argument.